### PR TITLE
Fix filename completions on a nonexisting file outside your home directory

### PR DESCRIPTION
**Pull request by ujihisa**
## Problem

ddc-file does not provide filenames when all of the following conditions have met:

* vim8
* not at a normal file (e.g. deol-edit buffer)
* not at your home directory

https://asciinema.org/a/rIvUFnqEAJhzm5stMSLu1KUUM

## Solution

This patch is not beautiful at all but it solves the problem.

https://asciinema.org/a/J3F5dGk85AOY8m6rxN5A1cVJE

### CODE REVIEW

File: autoload/ddc_file/internal.vim [hunk 1]
```diff
@@ -26,7 +26,13 @@ endfunction
 function! ddc_file#internal#info(input_line, is_posix) abort
   let input_file_full = s:line_to_file_full(a:input_line, a:is_posix)
   let input_file_base_prefix = s:full_to_base_prefix(input_file_full, a:is_posix)
-  let buf_path = expand('%:p')
+
+  if &buftype ==# '' && expand('%:p') !=# ''
```
Comment (Shougo) [inline autoload/ddc_file/internal.vim:30]: I think it should be `if &buftype ==# '' && !isdirectory(expand('%:p:h'))`

**Comment by Shougo** [comment]
`buf_path` is used to detect buffer directory.  So `buf_path` is invalid, you cannot complete path.

**Comment by ujihisa** [comment]
Added 7bc210550eef778edbd20c6730b046f668e06610 to fix the known issue described in the description, and updated the description itself!

Removed part:
 
> ## Known issue
> 
> It keeps the files at the previous directory too...

**Comment by Shougo** [comment]
```diff
diff --git a/denops/@ddc-sources/file.ts b/denops/@ddc-sources/file.ts
index 47efd4e..c55e819 100644
--- a/denops/@ddc-sources/file.ts
+++ b/denops/@ddc-sources/file.ts
@@ -2,6 +2,7 @@ import {
   BaseSource,
   Candidate,
   GatherCandidatesArguments,
+  fn,
   homeDir,
   path as univPath,
   wrapA,
@@ -139,7 +140,7 @@ export class Source extends BaseSource<Params> {

     // point from cwd
     findPointsAsync.push({
-      dir: Deno.cwd(),
+      dir: await fn.getcwd(args.denops) as string,
       max: p.cwdMaxCandidates,
       menu: p.displayCwd,
       asRoot: p.cwdAsRoot,
```

This is better.

**Comment by ujihisa** [comment]
@Shougo Thx! Added 289fc133e3c7e9d004f60c8bd2c38b950825fc26 to use its getcwd.

**Comment by ujihisa** [comment]
Oops

**Comment by LumaKernel** [comment]
Thank you! Formatting looks failing. Can you run `deno fmt`?

**Comment by LumaKernel** [comment]
https://github.com/LumaKernel/ddc-file/blob/6ea53360935eceecd2882389a70c0381bdbaabdc/denops/%40ddc-sources/file.ts#L152

Could you please fix here too with storing evaluated cwd into variable?

**Comment by ujihisa** [comment]
Thx! Fixed both of them.

**Comment by LumaKernel** [comment]
Merged. Much appreciated!

**Comment by ujihisa** [comment]
Thx!

**Pull request merged**